Mathlib/Util/AssertExistsExt.lean

```diff
@@ -0,0 +1,55 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Lean.Environment
+
+/-!
+# Environment extension for tracking existence of declarations and imports
+
+This is used by the `assert_not_exists` and `assert_not_imported` commands.
+-/
+
+section
+open Lean Elab Meta
+
+namespace Mathlib.AssertNotExist
+
+/-- `AssertExists` is the structure that carries the data to check if a declaration or an
+import are meant to exist somewhere in Mathlib. -/
+structure AssertExists where
+  /-- The type of the assertion: `true` means declaration, `false` means import. -/
+  isDecl : Bool
+  /-- The fully qualified name of a declaration that is expected to exist. -/
+  givenName : Name
+  /-- The name of the module where the assertion was made. -/
+  modName : Name
+  deriving BEq, Hashable
+
+/-- Defines the `assertExistsExt` extension for adding a `HashSet` of `AssertExists`s
+to the environment. -/
+initialize assertExistsExt : SimplePersistentEnvExtension AssertExists (HashSet AssertExists) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := fun as => as.foldl HashSet.insertMany {}
+    addEntryFn := .insert
+  }
+
+/--
+`addDeclEntry isDecl declName mod` takes as input the `Bool`ean `isDecl` and the `Name`s of
+a declaration or import, `declName`, and of a module, `mod`.
+It extends the `AssertExists` environment extension with the data `isDecl, declName, mod`.
+This information is used to capture declarations and modules that are required to not
+exist/be imported at some point, but should eventually exist/be imported.
+-/
+def addDeclEntry {m : Type → Type} [MonadEnv m] (isDecl : Bool) (declName mod : Name) : m Unit :=
+  modifyEnv (assertExistsExt.addEntry · { isDecl := isDecl, givenName := declName, modName := mod })
+
+/-- `getSortedAssertExists env` returns the array of `AssertExists`, placing first all declarations,
+in alphabetical order, and then all modules, also in alphabetical order. -/
+def getSortedAssertExists (env : Environment) : Array AssertExists :=
```

Comment:
Do you want this to live in the `Lean.Environment` namespace for access to dot notation? (This seems standard.) 

Optional suggestion: sorting could be a boolean parameter you pass. Either is fine to me.

Reply:
I put `getSortedAssertExists` in the `Lean.Environment` namespace.

I did not do the optional sorting, since the `HashSet`s produce random orderings and for both `#guard_msgs` and retrieving information about a specific declaration/import it is convenient to know where to look.